cli/src/main/java/com/devonfw/tools/ide/cli/IdeCompleter.java

```diff
@@ -0,0 +1,217 @@
+        wordsWithoutOptions.add(singleWord);
+      }
+    }
+
+    if (word.startsWith("-") && wordsWithoutOptions.isEmpty()) {
+      addCandidates(candidates, cleanedOptions); // adds rest of options without used option
+    }
+
+    if (wordsWithoutOptions.size() == 1) {
+      addCandidates(candidates, this.commandlets); // adds all commandlets
+    } else if (wordsWithoutOptions.size() == 2) {
+      // 2nd layer..
+
+      Commandlet commandlet = context.getCommandletManager().getCommandlet(wordsWithoutOptions.get(0));
+      if (commandlet != null) {
+        properties = commandlet.getProperties();
+        for (Property<?> property : properties) {
+          if (property instanceof ToolProperty) {
+            addCandidates(candidates, this.toolCommandlets);
+          }
+          if (commandlet instanceof HelpCommandlet) { // help completion
+            Set<String> commandletsWithoutHelp;
+            commandletsWithoutHelp = this.commandlets;
+            commandletsWithoutHelp.remove(commandlet.getName());
+            addCandidates(candidates, commandletsWithoutHelp);
+          }
+        }
+      }
+      // 3rd layer
+    } else if (wordsWithoutOptions.size() == 3) {
+      Commandlet commandlet = context.getCommandletManager().getCommandlet(wordsWithoutOptions.get(0));
+      if (commandlet != null) {
+        properties = commandlet.getProperties();
+        for (Property<?> property : properties) {
+          if (property instanceof VersionProperty) { // add version numbers
+            Commandlet subCommandlet = context.getCommandletManager().getCommandlet(wordsWithoutOptions.get(1));
+            if (subCommandlet != null) {
+              String toolEdition = context.getVariables().getToolEdition(subCommandlet.getName());
+              List<VersionIdentifier> versions = context.getUrls().getSortedVersions(subCommandlet.getName(),
+                  toolEdition);
+              int sort = 0;
+              // adds version numbers in sorted order (descending)
+              for (VersionIdentifier vi : versions) {
+                sort++;
+                String versionName = vi.toString();
+                candidates.add(new Candidate(versionName, versionName, null, null, null, null, true, sort));
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private boolean checkInvalidPattern(String commandLine) {
+
+    Matcher matcher = patternCompiled.matcher(commandLine);
+
+    return matcher.find();
+  }
+
+  private void addCandidates(List<Candidate> candidates, Iterable<String> cands) {
+
+    addCandidates(candidates, cands, "", "", true);
+  }
+
+  private void addCandidates(List<Candidate> candidates, Iterable<String> cands, String preFix, String postFix,
+      boolean complete) {
+
+    for (String s : cands) {
+      candidates
+          .add(new Candidate(AttributedString.stripAnsi(preFix + s + postFix), s, null, null, null, null, complete));
+    }
+  }
```

Comment:
I do not see any invocation using the method directly so why having `preFix` and `postFix`feature if we never use it?

Reply:
Was taken from the picocli implementation example could be removed I think.